### PR TITLE
[Spark load][Fe 5/6] Fe submit spark etl job

### PR DESCRIPTION
**Pull request by wyb**
After user creates a spark load job which status is PENDING, Fe will schedule and submit the spark etl job.
1. Begin transaction
2. Create a SparkLoadPendingTask for submitting etl job
2.1 Create etl job configuration according to https://github.com/apache/incubator-doris/issues/3010#issuecomment-635174675
2.2 Upload the configuration file and job jar to HDFS with broker
2.3 Submit etl job to spark cluster
2.4 Wait for etl job submission result
3. Update job state to ETL and log job update info if etl job is submitted successfully

#3433

### CODE REVIEW

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/catalog/OlapTable.java [hunk 1]
```diff
@@ -530,6 +530,30 @@ public KeysType getKeysType() {
         return keysType;
     }
 
+    public KeysType getKeysTypeByIndexId(long indexId) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/OlapTable.java:533]: There is `keysType` is `MaterializedIndexMeta`. You can get it directly.

File: fe/src/main/java/org/apache/doris/common/Pair.java [hunk 2]
```diff
@@ -25,7 +27,9 @@
 public class Pair<F, S> {
     public static PairComparator<Pair<?, Comparable>> PAIR_VALUE_COMPARATOR = new PairComparator<>();
 
+    @SerializedName(value = "first")
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/common/Pair.java:30]: I'am not sure this is ok, cause there is no guarantee that the `F` and `S` object can also be serialized by GSON

File: fe/src/main/java/org/apache/doris/common/util/BrokerUtil.java [hunk 3]
```diff
@@ -17,52 +17,59 @@
...
+import java.io.UnsupportedEncodingException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
 import java.util.Collections;
 import java.util.List;
 
 public class BrokerUtil {
     private static final Logger LOG = LogManager.getLogger(BrokerUtil.class);
 
+    private static int READ_BUFFER_SIZE = 1024 * 1024;
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/common/util/BrokerUtil.java:67]: add unit to name.

File: fe/src/main/java/org/apache/doris/load/loadv2/SparkEtlJobHandler.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+ * 5. delete etl output path
+ */
+public class SparkEtlJobHandler {
+    private static final Logger LOG = LogManager.getLogger(SparkEtlJobHandler.class);
+
+    private static final String APP_RESOURCE_NAME = "palo-fe.jar";
+    private static final String CONFIG_FILE_NAME = "jobconfig.json";
+    private static final String APP_RESOURCE_LOCAL_PATH = PaloFe.DORIS_HOME_DIR + "/lib/" + APP_RESOURCE_NAME;
+    private static final String JOB_CONFIG_DIR = "configs";
+    private static final String MAIN_CLASS = "org.apache.doris.load.loadv2.etl.SparkEtlJob";
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/load/loadv2/SparkEtlJobHandler.java:55]: How about get it from `SparkEtlJob.class.getXXX()`?

File: fe/src/main/java/org/apache/doris/load/loadv2/SparkEtlJobHandler.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                break;
+            }
+
+            // check state and retry
+            state = handle.getState();
+            if (fromSparkState(state) == TEtlState.CANCELLED) {
+                throw new LoadException(errMsg + "spark app state: " + state.toString());
+            }
+            if (retry >= GET_APPID_MAX_RETRY_TIMES) {
+                throw new LoadException(errMsg + "wait too much time for getting appid. spark app state: "
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/load/loadv2/SparkEtlJobHandler.java:127]:
```suggestion
                throw new LoadException(errMsg + " wait too much time for getting appi d. spark app state: "
```

**Review by morningman — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/common/Pair.java (hunk 2 above)
Comment (wyb) [reply fe/src/main/java/org/apache/doris/common/Pair.java:30]: Users guarantee this when use Pair class？like Map and List.

File: fe/src/main/java/org/apache/doris/load/loadv2/SparkEtlJobHandler.java (hunk 4 above)
Comment (wyb) [reply fe/src/main/java/org/apache/doris/load/loadv2/SparkEtlJobHandler.java:55]: ok, I comment and will replace with it when SparkEtlJob class is merged.

File: fe/src/main/java/org/apache/doris/load/loadv2/SparkEtlJobHandler.java (hunk 5 above)
Comment (wyb) [reply fe/src/main/java/org/apache/doris/load/loadv2/SparkEtlJobHandler.java:130]: errMsg already have a space at the end.

File: fe/src/main/java/org/apache/doris/common/util/BrokerUtil.java [hunk 6]
```diff
@@ -139,4 +148,352 @@ public static String printBroker(String brokerName, TNetworkAddress address) {
...
+            } catch (TException e) {
+                reopenClient(client);
+                tReadResponse = client.pread(tPReadRequest);
+            }
+            if (tReadResponse.getOpStatus().getStatusCode() != TBrokerOperationStatusCode.OK) {
+                throw new UserException("Broker read failed. path=" + path + ", broker=" + address
+                                                + ", msg=" + tReadResponse.getOpStatus().getMessage());
+            }
+            failed = false;
+            return tReadResponse.getData();
```
[the remark this answers is not in the document]
Comment (wyb) [reply fe/src/main/java/org/apache/doris/common/util/BrokerUtil.java:219]: ok

File: fe/src/main/java/org/apache/doris/common/Pair.java (hunk 2 above)
Comment (wyb) [reply fe/src/main/java/org/apache/doris/common/Pair.java:32]: I add a comment.
When using Pair for persistence, users need to guarantee that F and S can be serialized through Gson

**Review by morningman — DISMISSED** [review]
LGTM

**Review by imay — DISMISSED** [review]

**Review by imay — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**